shadow/state.go

```diff
@@ -180,3 +187,121 @@ func updateState(state map[string]interface{}, update map[string]interface{}) er
 func hasUpdate(s json.RawMessage) bool {
 	return len(s) != 0
 }
+
+// NestedState is JSON unmarshaller for state metadata.
+type NestedState map[string]interface{}
+
+func (n *NestedState) UnmarshalJSON(b []byte) error {
+	if *n == nil {
+		*n = make(map[string]interface{})
+	}
+	j := make(map[string]json.RawMessage)
+	if err := json.Unmarshal(b, &j); err != nil {
+		return err
+	}
+	for k, v := range j {
+		v2, err := unmarshalStateImpl(v)
+		if err != nil {
+			return err
+		}
+		(*n)[k] = v2
+	}
+	return nil
+}
+
+func unmarshalStateImpl(b []byte) (interface{}, error) {
+	switch {
+	case b[0] == '[' && b[len(b)-1] == ']':
+		var v2 []json.RawMessage
+		if err := json.Unmarshal(b, &v2); err != nil {
+			return nil, err
+		}
+		var v5 []interface{}
+		for _, v3 := range v2 {
+			v4, err := unmarshalStateImpl(v3)
+			if err != nil {
+				return nil, err
+			}
+			v5 = append(v5, v4)
+		}
+		return v5, nil
+	case b[0] == '{':
+		var v2 NestedState
+		if err := json.Unmarshal(b, &v2); err != nil {
+			return nil, err
+		}
+		return v2, nil
+	default:
+		var v2 interface{}
+		if err := json.Unmarshal(b, &v2); err != nil {
+			return nil, err
+		}
+		return v2, nil
+	}
+}
+
+// MapTo copies NestedState to the given typed struct.
+func (n NestedState) MapTo(dst interface{}) error {
+	return n.mapToImpl(reflect.ValueOf(dst).Elem())
```

Comment:
[mitchellh/mapstructure](https://github.com/mitchellh/mapstructure) can convert a `map[string]interface` to a struct and it's used widely and tested well for the purpose, so I think it's better to use the library here, or we might not have to provide such feature on this SDK (since users can convert a map by any way they want like mitchellh/mapstructure)

What do you think?

Reply:
Using mapstructure in the shadow example sounds clean.